### PR TITLE
Playground autocomplete

### LINKED ISSUE #78: Playground- autocomplete
Autocomplete Sass-specific keywords (e.g. `@use`, `@debug`) and in-scope variables

### PR DESCRIPTION
**Pull request by jamesnw**
## Description
Adds Sass-specific Autocomplete

## Related Issue(s)
#78 

## Steps to test/reproduce
Go to [Playground](https://deploy-preview-77--sass-site-oddbird.netlify.app/playground/), type in Sass-specific words.

1. At-rules, like `@use`, `@debug`, etc.
2. Built in module names- `@use "sass:math"`, for instance
3. Built in module variables only if module is imported- `math.$pi`
4. Built in module functions only if module is imported- `math.clamp`

- [x] Miriam check functionality
- [x] Stacy check functionality
- [x] Jonny review code

### CODE REVIEW

**Comment by jamesnw** [comment]
@mirisuzanne  and @stacyk - could you try out the Autocomplete [here](https://deploy-preview-77--sass-site-oddbird.netlify.app/playground/) and see if it works well and as expected?

**Review by mirisuzanne — APPROVED** [review]
1. At-rules: works great
2. Built in module names: only works with e.g. `"sass:math"`, not `'sass:math'`. Should probably work with single-quotes as well?
3. Built in variables: works great
4. Built in functions: works great

**Comment by jamesnw** [comment]
> 2\. Built in module names: only works with e.g. `"sass:math"`, not `'sass:math'`. Should probably work with single-quotes as well?

Good catch! Fixed.

File: source/assets/js/playground/autocomplete.ts [hunk 1]
```diff
@@ -0,0 +1,297 @@
...
+  'error',
+  'warn',
+  'debug',
+  'at-root',
+  'if',
+  'else',
+  'each',
+  'for',
+  'while',
+];
```
Comment (jgerigmeyer) [inline source/assets/js/playground/autocomplete.ts:28]: Do we want to add `@media`, `@supports`, and `@keyframes` here? And maybe also add autocomplete for other CSS at-rules?

File: source/assets/js/playground/autocomplete.ts [hunk 2]
```diff
@@ -0,0 +1,297 @@
...
+};
+
+type ModuleDefinition = CompletionInfo & {
+  functions?: CompletionInfo[];
+  variables?: CompletionInfo[];
+};
+
+const builtinModules: ModuleDefinition[] = [
+  {
+    name: 'color',
```
Comment (jgerigmeyer) [inline source/assets/js/playground/autocomplete.ts:60]: Is there a reason we're not including the `color` functions (e.g. `adjust`, `scale`, etc) here?

File: source/assets/js/playground/autocomplete.ts [hunk 3]
```diff
@@ -0,0 +1,297 @@
...
+      {name: 'has-key'},
+      {name: 'deep-merge'},
+      {name: 'deep-remove'},
+    ],
+  },
+  {
+    name: 'math',
+    description: 'provides functions that operate on numbers',
+    variables: [
+      {name: '$e', description: undefined},
```
Comment (jgerigmeyer) [inline source/assets/js/playground/autocomplete.ts]: Why set `description: undefined`?

File: source/assets/js/playground/autocomplete.ts [hunk 4]
```diff
@@ -0,0 +1,297 @@
...
+    ],
+  },
+  {
+    name: 'meta',
+    description: 'exposes the details of Sass’s inner workings',
+    functions: [
+      {name: 'apply'},
+      {name: 'load-css'},
+      {name: 'accepts-content'},
+      {name: 'calc-args'},
```
Comment (jgerigmeyer) [inline source/assets/js/playground/autocomplete.ts:143]: Missing `calc-name`?

```suggestion
      {name: 'calc-args'},
      {name: 'calc-name'},
```

File: source/assets/js/playground/autocomplete.ts [hunk 5]
```diff
@@ -0,0 +1,297 @@
...
+      {name: 'mixin-exists'},
+      {name: 'variable-exists'},
+      {name: 'get-function'},
+      {name: 'get-mixin'},
+      {name: 'global-variable-exists'},
+      {name: 'keywords'},
+      {name: 'module-functions'},
+      {name: 'module-mixins'},
+      {name: 'module-variables'},
+      {name: 'type-of'},
```
Comment (jgerigmeyer) [inline source/assets/js/playground/autocomplete.ts:158]: Missing `inspect`?

```suggestion
      {name: 'type-of'},
      {name: 'inspect'},
```

File: source/assets/js/playground/autocomplete.ts [hunk 6]
```diff
@@ -0,0 +1,297 @@
...
+    ],
+  },
+  {
+    name: 'selector',
+    description: 'provides access to Sass’s powerful selector engine',
+    functions: [
+      {name: 'isSuperselector'},
+      {name: 'simpleSelectors'},
```
Comment (jgerigmeyer) [inline source/assets/js/playground/autocomplete.ts]:
```suggestion
      {name: 'is-superselector'},
      {name: 'simple-selectors'},
```

File: source/assets/js/playground/autocomplete.ts [hunk 7]
```diff
@@ -0,0 +1,297 @@
...
+      {name: 'index'},
+      {name: 'slice'},
+      {name: 'unique-id'},
+      {name: 'split'},
+    ],
+  },
+];
+
+const moduleNames = builtinModules.map(mod => mod.name);
+const moduleNameRegExp = new RegExp(`(${moduleNames.join('|')}).\\$?(\\w)*`);
```
Comment (jgerigmeyer) [inline source/assets/js/playground/autocomplete.ts]: Are the captured groups used here? If not, what's the reason to capture the final `(\w)*` group?

**Review by jgerigmeyer — COMMENTED** [review]
@jamesnw Works great! Obviously it would be ideal if we could get some of these values (e.g. at-rules, modules, etc) dynamically from Sass itself. I assume you looked into that and they're not being exposed currently? It might even be worth mentioning that to Natalie to see what she thinks.

File: source/assets/js/playground/autocomplete.ts (hunk 1 above)
Comment (mirisuzanne) [reply source/assets/js/playground/autocomplete.ts:28]: I think css at-rules would be good to consider supporting. Hopefully there's a clear enough division between that and auto-completing css properties/values, so this doesn't become a very slippery scope slope.

Comment (jamesnw) [reply source/assets/js/playground/autocomplete.ts:28]: I think that probably should go upstream into [@codemirror/lang-css](https://github.com/codemirror/lang-css/blob/main/src/complete.ts)- that shouldn't be too challenging, I don't think.

We are already getting auto-completion of css properties and values from that package.

File: source/assets/js/playground/autocomplete.ts (hunk 3 above)
Comment (jamesnw) [reply source/assets/js/playground/autocomplete.ts]: It was for a type issue that doesn't exist anymore, removed.

**Comment by stacyk** [comment]
@jamesnw Will this story (or any future stories) include autocompleting CSS properties? For instance, in the SCSS panel, if I type 

```
.foo {
  di
 ```
 
 I was expecting to see `display` as an option but instead I only see html elements
 
Screenshot 2024-09-05 at 12 27 18 PM

**Comment by jamesnw** [comment]
> @jamesnw Will this story (or any future stories) include autocompleting CSS properties? For instance, in the SCSS panel, if I type
> 
> ```
> .foo {
>   di
> ```
> 
> I was expecting to see `display` as an option but instead I only see html elements

The issue is the missing `;` on the first line is messing with the parser, and it can't determine the context-

image

**Comment by stacyk** [comment]
@jamesnw I'm not sure the full scope of this story or if it is even unexpected, but I noticed when I type `@use` and select `sass:math` but realize a second later I wanted `sass:maps`, if I delete a few characters from the end it won't give me the same options. After doing a bit more testing I see I can't delete that final quote. Is that the expected user behavior? I thought it should auto complete with sass:* suggestions without that final quote.

File: source/assets/js/playground/autocomplete.ts (hunk 7 above)
Comment (jamesnw) [reply source/assets/js/playground/autocomplete.ts]: Good catch, removed.

**Comment by stacyk** [comment]
> > @jamesnw Will this story (or any future stories) include autocompleting CSS properties? For instance, in the SCSS panel, if I type
> > ```
> > .foo {
> >   di
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I was expecting to see `display` as an option but instead I only see html elements
> 
> The issue is the missing `;` on the first line is messing with the parser, and it can't determine the context-
> 
> image

I'm literally LOL

**Comment by jamesnw** [comment]
> @jamesnw I'm not sure the full scope of this story or if it is even unexpected, but I noticed when I type `@use` and select `sass:math` but realize a second later I wanted `sass:maps`, if I delete a few characters from the end it won't give me the same options. After doing a bit more testing I see I can't delete that final quote. Is that the expected user behavior? I thought it should auto complete with sass:* suggestions without that final quote.

This is another instance of confusing the parser. Without the end quote, the parser doesn't know it's a StringLiteral, and thinks it's a pseudoclass. I pushed a change that should address it.

**Review by stacyk — APPROVED** [review]
The update you made for the autocomplete without the final quote works great!

File: source/assets/js/playground/autocomplete.ts (hunk 2 above)
Comment (jamesnw) [reply source/assets/js/playground/autocomplete.ts:60]: I'm holding out hope there's a way to get this from Sass directly, as this is a long list, and will be changing once the CSS Color 4 changes are made in Sass.

File: source/assets/js/playground/editor-setup.ts [hunk 8]
```diff
@@ -11,7 +11,7 @@ import {
   historyKeymap,
   indentWithTab,
 } from '@codemirror/commands';
-import {css as langCss} from '@codemirror/lang-css';
+import {css as langCss, cssCompletionSource} from '@codemirror/lang-css';
```
Comment (jgerigmeyer) [inline source/assets/js/playground/editor-setup.ts]: Unused import?

```suggestion
import {css as langCss} from '@codemirror/lang-css';
```

File: source/assets/js/playground/autocomplete.ts (hunk 1 above)
Comment (jgerigmeyer) [reply source/assets/js/playground/autocomplete.ts:28]: Makes sense -- add a comment or file a new issue so we don't lose track?

File: source/assets/js/playground/autocomplete.ts (hunk 2 above)
Comment (jgerigmeyer) [reply source/assets/js/playground/autocomplete.ts:60]: Makes sense -- add a comment so we don't forget?

**Review by jgerigmeyer — APPROVED** [review]
👏

**Review by jgerigmeyer — APPROVED** [review]

**Comment by jamesnw** [comment]
Moving to sass repo- https://github.com/sass/sass-site/pull/1166

**Pull request closed without merging**